### PR TITLE
create products categories

### PR DESCRIPTION
**Pull request by RosaNguyen158**
- Create products categories
- Closure categories
- Relationship between products & categories

### CODE REVIEW

**Review by zgid123 — CHANGES_REQUESTED** [review]

File: workspaces/api-gateway/src/typeOrm.config.ts [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
+
 import type { DataSourceOptions } from 'typeorm';
 import type { TypeOrmModuleOptions } from '@nestjs/typeorm';
 
 export function createTypeOrmOptions(): TypeOrmModuleOptions {
   return {
     migrationsTableName: 'migrations',
     type: 'postgres',
+    logger: 'simple-console',
+    logging: 'all',
```
Comment (zgid123) [inline workspaces/api-gateway/src/typeOrm.config.ts]: only do log if dev env

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        productId: product.id,
+        categoryId,
+      });
+    }
+
+    return this.productRepository.findOne({
+      where: {
+        slug,
+      },
+    });
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: why we need to `findOne` here? There should be a way to update the product with repository

File: workspaces/api-gateway/src/api/products/services/GetListProductsService.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Like, Repository } from 'typeorm';
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/GetListProductsService.ts]: Repository is a type no?

File: workspaces/api-gateway/src/api/products/services/GetListProductsService.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  constructor(
+    @InjectRepository(Product)
+    private productRepository: Repository<Product>,
+  ) {}
+
+  public async exec({
+    name,
+    page,
+    perpage,
+  }: PagyProductsDto): Promise<ProductsPagyMetadataVM> {
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/GetListProductsService.ts]: return type of service should not be a view model. Must use return type of `findAndCount`

File: workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  ): Promise<ProductsCategories[]> {
+    const { categoryId, productId } = createProCateDto;
+
+    const listCategories = await this.getAncestorCategoriesService.exec(
+      categoryId,
+    );
+
+    const listIdCategories = getListIdCategories(listCategories);
+
+    const newProCate = listIdCategories.map((idCategory: string) =>
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts]: should be `newProCats`

File: workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  public async exec(
+    createProCateDto: CreateProCateDto,
+  ): Promise<ProductsCategories[]> {
+    const { categoryId, productId } = createProCateDto;
+
+    const listCategories = await this.getAncestorCategoriesService.exec(
+      categoryId,
+    );
+
+    const listIdCategories = getListIdCategories(listCategories);
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts]: should be `listCategoryIds`

File: workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+      const listIdCategories = getListIdCategories(listCategories);
+
+      const level = listIdCategories.length;
+
+      if (category.parentId !== null) {
+        await QueryRunner.manager.query(
+          `DELETE FROM categories_closure WHERE id_ancestor = '${category.parentId}' AND id_descendant <> '${category.parentId}'`,
+        );
+      }
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts]: why we need this one? https://typeorm.io/tree-entities#working-with-tree-entities didn't this one work?

File: workspaces/api-gateway/src/__tests__/api/auth/GetProfileService.spec.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  beforeEach(async () => {
+    const salt = await genSalt();
+    const hashedPassword = await hash(userDto.password, salt);
+    user = userRepository.create({
+      username: userDto.username,
+      password: hashedPassword,
+    });
+    await userRepository.save(user);
+  });
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/auth/GetProfileService.spec.ts]: should be in `beforeAll`

**Review by zgid123 — CHANGES_REQUESTED** [review]

File: workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts (hunk 7 above)
Comment (RosaNguyen158) [reply workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts]: Em tìm trong link trên không thấy có hỗ trợ delete closure, với categories_closure không tạo entity nên không dùng QueryBuilder của typeorm được ạ

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts (hunk 2 above)
Comment (RosaNguyen158) [reply workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: update trả về giá trị affect là 0 hoặc 1, em muốn nó trả về giá trị của product sau cập nhật nên dùng findOne ạ

Comment (zgid123) [reply workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: https://typeorm.io/repository-api#repository-api #save section

File: workspaces/api-gateway/src/api/products/services/GetListProductsService.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  constructor(
+    @InjectRepository(Product)
+    private productRepository: Repository<Product>,
+  ) {}
+
+  public async exec({
+    name,
+    page,
+    perpage,
+  }: PagyProductsDto): Promise<{ data: Product[]; total: number }> {
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/GetListProductsService.ts]: why dont you use the return type of `findAndCount`?

File: workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+      await this.categoryRepository.update(category.id, {
+        name,
+        slug,
+        parentId,
+        level,
+        ...data,
+      });
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts]: why this one duplicate with the else block?

File: workspaces/api-gateway/src/api/categories/services/GetAncestorCategoriesService.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  constructor(
+    @InjectRepository(Category)
+    private categoryRepository: Repository<Category>,
+    private getCategoryByKeyService: GetCategoryByKeyService,
+  ) {}
+
+  public async exec(idCategory: string): Promise<Category> {
+    const category = await this.getCategoryByKeyService.exec(idCategory);
+
+    const findAncestors = await this.categoryRepository.manager
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/GetAncestorCategoriesService.ts]: can we merge this service with the `GetDescendantCategoryService`?

File: workspaces/api-gateway/src/api/categories/services/CreateCategoryService.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      name,
+      slug: await createSlug(
+        {
+          name,
+          columnName: 'slug',
+        },
+        this.checkedCategoryExistedService.exec.bind(this),
+      ),
+      level: level,
+      parentId: parentId,
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/CreateCategoryService.ts]: eslint should warn this one. If the key has the same name of the variable, no need to pass the variable. Must check the eslint rule

File: workspaces/api-gateway/src/__tests__/api/auth/GetProfileService.spec.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+  afterEach(async () => {
+    await userRepository.clear();
+  });
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/auth/GetProfileService.spec.ts:41]: why we need this one?

File: workspaces/api-gateway/src/__tests__/api/auth/SignInService.spec.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+  beforeEach(async () => {
+    const salt = await genSalt();
+    const hashedPassword = await hash(userDto.password, salt);
+    user = userRepository.create({
+      username: userDto.username,
+      password: hashedPassword,
+    });
+    await userRepository.save(user);
+  });
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/auth/SignInService.spec.ts]: every data creation must be in `beforeAll` and every data clear must be in `afterAll`

**Review by zgid123 — CHANGES_REQUESTED** [review]

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import type { Repository } from 'typeorm';
+
+import { Product } from 'database/models/product.entity';
+import { ProductsCategories } from 'database/models/productsCategories.entity';
+import { CheckProductExistedService } from 'api/products/services/CheckProductExistedService';
+import { CreateProductCategoryService } from 'api/products/services/CreateProductCategoryService';
+import { createSlug } from '../../../utils';
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: why don't we use absolute path here?

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    });
+
+    if (categoryId) {
+      await this.productCategoryRepository.delete({ productId: product.id });
+      await this.createProductCategoryService.exec({
+        productId: product.id,
+        categoryId,
+      });
+    }
+
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: what happens if the categoryId is passed and its value is the same?

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+      );
+    }
+
+    product = await this.productRepository.save({
+      id: product.id,
+      name,
+      slug,
+      categoryId,
+      ...data,
+    });
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: must do

```ts
Object.assign(product, {
  name,
  slug,
  categoryId,
  ...data,
});

await this.productRepository.save(product);
```

File: workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+      key: categoryId,
+      descendant: false,
+    });
+
+    const listCategoryIds = getlistCategoryIds(listCategories);
+
+    const newProCats = listCategoryIds.map((idCategory: string) =>
+      this.proCateRepository.create({
+        productId: productId,
+        categoryId: idCategory,
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts]: the variable name must be `categoryId`

File: workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    const listCategories = await this.getRelatedCategoriesService.exec({
+      key: categoryId,
+      descendant: false,
+    });
+
+    const listCategoryIds = getlistCategoryIds(listCategories);
+
+    const newProCats = listCategoryIds.map((idCategory: string) =>
+      this.proCateRepository.create({
+        productId: productId,
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/services/CreateProductCategoryService.ts]: as I mentioned, if the key is same as variable, do not need the variable here

File: workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts (hunk 10 above)
Comment (RosaNguyen158) [reply workspaces/api-gateway/src/api/categories/services/UpdateCategoryService.ts]: do update có 2 trường hợp, trên if là update parentId, sau khi update bên bảng products xong mới update bảng products categories được ạ, còn dưới chỗ else thì chỉ cần update bên products thôi k cần tính toán nhiều nên em để riêng hai blocks á anh

File: workspaces/api-gateway/src/api/products/product.controller.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  @Post()
+  @UseInterceptors(MapInterceptor(Product, ProductVM))
+  public async create(
+    @Body() createProductDto: CreateProductDto,
+  ): Promise<Product> {
+    return this.createProductService.exec(createProductDto);
+  }
+
+  @Get()
+  @UseInterceptors(MapInterceptor(Array, ProductsPagyMetadataVM))
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/products/product.controller.ts]: why we do mapping with Array here?

File: workspaces/api-gateway/src/api/categories/services/GetRelativeCategoryService.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+
+    if (descendant) {
+      relatedCategories = await this.categoryRepository.manager
+        .getTreeRepository(Category)
+        .findDescendantsTree(category);
+    } else {
+      relatedCategories = await this.categoryRepository.manager
+        .getTreeRepository(Category)
+        .findAncestorsTree(category);
+    }
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/GetRelativeCategoryService.ts]: this code block should be refactored

File: workspaces/api-gateway/src/api/categories/services/GetListCategoriesService.ts [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+  public async exec({
+    name,
+    page,
+    perpage,
+  }: PagyCategoryDto): Promise<[Category[], number]> {
+    return this.categoryRepository.findAndCount({
+      where: { name: Like(`%${name}%`) },
+      order: { name: 'DESC' },
+      take: perpage,
+      skip: (page - 1) * perpage,
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/categories/services/GetListCategoriesService.ts]: what if they pass the 0 here?

File: workspaces/api-gateway/src/__tests__/api/products/products.controller.spec.ts [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+  describe('POST /api/products', () => {
+    describe('with valid product name', () => {
+      it('returns product', async () => {
+        const resultMapping = await mapper.mapAsync(
+          product,
+          Product,
+          ProductVM,
+        );
+
+        jest.spyOn(Date.prototype, 'getTime').mockImplementation(() => 123);
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/products/products.controller.spec.ts]: this mock is used on multiple places, should include this one inside `createModule` helper

File: workspaces/api-gateway/src/__tests__/api/products/products.controller.spec.ts [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+  });
+
+  afterEach(async () => {
+    const QueryRunner = dataSource.createQueryRunner();
+
+    await QueryRunner.manager.query('TRUNCATE products_categories CASCADE');
+    await QueryRunner.manager.query('TRUNCATE products CASCADE');
+    await QueryRunner.manager.query('TRUNCATE categories_closure CASCADE');
+    await QueryRunner.manager.query('TRUNCATE categories CASCADE');
+  });
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/products/products.controller.spec.ts]: why we need to do those things for each test case?

**Review by zgid123 — CHANGES_REQUESTED** [review]

File: workspaces/api-gateway/src/api/categories/services/GetAncestorCategoriesService.ts (hunk 11 above)
Comment (RosaNguyen158) [reply workspaces/api-gateway/src/api/categories/services/GetAncestorCategoriesService.ts]: em merge 2 hàm đó lại 1 cái rồi ạ

File: workspaces/api-gateway/src/api/products/services/UpdateProductService.ts (hunk 15 above)
Comment (RosaNguyen158) [reply workspaces/api-gateway/src/api/products/services/UpdateProductService.ts]: chỗ này ban đầu em dùng absolute path mà lúc chạy test bị lỗi nên em phải đổi qua relative path mới chạy được ạ

File: workspaces/api-gateway/src/database/models/user.entity.ts [hunk 25]
```diff
@@ -24,9 +17,7 @@ export class User {
...
   email: string;
 
-  @CreateDateColumn()
-  createdAt: Date;
-
-  @UpdateDateColumn()
-  updatedAt: Date;
+  @AutoMap()
+  @Column()
+  roles: Role;
```
Comment (zgid123) [inline workspaces/api-gateway/src/database/models/user.entity.ts]: one user has multiple roles?

File: workspaces/api-gateway/src/app.module.ts [hunk 26]
```diff
@@ -19,6 +22,8 @@ const routes: Routes = [authRoutes];
       strategyInitializer: classes(),
     }),
     AuthModule,
+    CategoriesModule,
+    ProductsModule,
```
Comment (zgid123) [inline workspaces/api-gateway/src/app.module.ts]: must be `AdminModule` to replace those two

**Review by zgid123 — CHANGES_REQUESTED** [review]

File: workspaces/api-gateway/src/database/models/user.entity.ts [hunk 27]
```diff
@@ -24,9 +17,7 @@ export class User {
...
   email: string;
 
-  @CreateDateColumn()
-  createdAt: Date;
-
-  @UpdateDateColumn()
-  updatedAt: Date;
+  @AutoMap()
+  @Column()
+  role: Role;
```
Comment (zgid123) [inline workspaces/api-gateway/src/database/models/user.entity.ts]: must use enum column https://typeorm.io/entities#enum-column-type

File: workspaces/api-gateway/src/api/auth/services/SignUpService.ts [hunk 28]
```diff
@@ -3,32 +3,36 @@ import {
...
+    password,
+    role,
+  }: AuthCredentialsDto): Promise<IAuthProps> {
     const payload: IJwtPayload = { username };
     const salt = await genSalt();
     const hashedPassword = await hash(password, salt);
     const user = this.userRepository.create({
       username: username,
       password: hashedPassword,
+      role: role || Role.User,
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/auth/services/SignUpService.ts]: sign up always role user

File: workspaces/api-gateway/src/api/auth/auth.vm.ts [hunk 29]
```diff
@@ -10,6 +12,9 @@ export class ProfileVM {
 
   @AutoMap()
   username: string;
+
+  @AutoMap()
+  roles: Role;
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/auth/auth.vm.ts]: why this is plural?

File: workspaces/api-gateway/src/__tests__/api/auth/signUpService.spec.ts [hunk 30]
```diff
@@ -4,22 +4,22 @@ import { JwtService } from '@nestjs/jwt';
...
 import type { AuthCredentialsDto } from 'api/auth/dto/auth-credential.dto';
 
-describe('AuthSerice', () => {
-  let authService: AuthService;
+describe('SignUpService', () => {
   let userRepository: Repository<User>;
   let dataSource: DataSource;
   let jwtService: JwtService;
-  let user: User;
+  let signUpService: SignUpService;
```
Comment (zgid123) [inline workspaces/api-gateway/src/__tests__/api/auth/signUpService.spec.ts]: why this file name start with lower case while other starts with upper case?

File: workspaces/api-gateway/src/api/admin/products/services/CreateProductService.ts [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+
+    const newProduct = this.productRepository.create({
+      name: name,
+      category: checkCategory,
+      slug: await createSlug(
+        {
+          name: name,
+          columnName: 'slug',
+        },
+        this.checkProductExistedService.exec.bind(this),
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/admin/products/services/CreateProductService.ts]: why bind `this` here? Try to remove the `private productRepository: Repository<Product>` in `checkProductExistedService` to see the result

File: workspaces/api-gateway/src/api/admin/products/services/CreateProductCategoryService.ts [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+    });
+
+    const listCategoryIds = getlistCategoryIds(listCategories);
+
+    const newProCats = listCategoryIds.map((categoryId: string) =>
+      this.proCateRepository.create({
+        productId,
+        categoryId,
+      }),
+    );
```
Comment (zgid123) [inline workspaces/api-gateway/src/api/admin/products/services/CreateProductCategoryService.ts]: why don't we use `bulkInsert`?

**Review by zgid123 — CHANGES_REQUESTED** [review]

**Review by zgid123 — APPROVED** [review]

**Pull request merged**